packages/IntegrationTest/windows/integrationtest/TestHostHarness.cpp

```diff
@@ -0,0 +1,138 @@
+  reactHost.InstanceSettings().RedBoxHandler(m_redboxHandler);
+  m_instanceLoadedRevoker = reactHost.InstanceSettings().InstanceLoaded(
+      winrt::auto_revoke, [weakThis{get_weak()}](const auto & /*sender*/, const auto &args) noexcept {
+        if (auto strongThis = weakThis.get()) {
+          strongThis->m_context = args.Context();
+          strongThis->StartListening();
+        }
+
+        ReactNotificationService notifications(args.Context().Notifications());
+
+        notifications.Subscribe(
+            TestModule::TestCompletedEvent(),
+            ReactDispatcher(args.Context().UIDispatcher()),
+            [weakThis](const auto &/*sender*/, ReactNotificationArgs<void> /*args*/) noexcept {
+              if (auto strongThis = weakThis.get()) {
+                strongThis->OnTestCompleted();
+              }
+            });
+
+        notifications.Subscribe(
+            TestModule::TestPassedEvent(),
+            ReactDispatcher(args.Context().UIDispatcher()),
+            [weakThis](const auto &/*sender*/, ReactNotificationArgs<bool> args) noexcept {
+              if (auto strongThis = weakThis.get()) {
+                strongThis->OnTestPassed(*args.Data());
+              }
+            });
+      });
+}
+
+winrt::fire_and_forget TestHostHarness::StartListening() noexcept {
+  auto listenResult = co_await m_commandListener->StartListening();
+
+  switch (listenResult) {
+    case ListenResult::Success:
+      co_return;
+
+    case ListenResult::AddressInUse:
+      ShowJSError("Test harness address is already in use. Is there another instance running?");
+      co_return;
+
+    case ListenResult::OtherError:
+      ShowJSError("The test harness ran into an unexpected error trying to listen for commands.");
+      co_return;
+
+    default:
+      ShowJSError("Unexpected ListenResult from TestCommandListener.");
+      co_return;
+  }
+}
+
+void TestHostHarness::OnTestCommand(const TestCommand &command, TestCommandResponse &&response) noexcept {
+  m_pendingResponse = std::move(response);
+
+  switch (command.Id) {
+    case TestCommandId::RunTestComponent: {
+      auto componentName = command.payload.GetString();
+      m_rootView.ComponentName(componentName);
+      break;
+    }
+
+    default: {
+      ShowJSError("Unexpected command ID from test runner");
+      m_pendingResponse->UnknownError("Unexpected command ID from test runner");
+      m_pendingResponse.reset();
+      break;
+    }
+  }
+}
+
+void TestHostHarness::OnTestCompleted() noexcept {
+  OnTestPassed(true);
+}
+
+void TestHostHarness::OnTestPassed(bool passed) noexcept {
+  if (m_pendingResponse) {
+    m_pendingResponse->TestPassed(passed);
+    m_pendingResponse.reset();
+  }
+}
+
+void TestHostHarness::ShowJSError(std::string_view err) noexcept {
+  m_context.CallJSFunction(L"RCTLog", L"logToConsole", "error", err);
+}
+
+void TestHostHarnessRedboxHandler::ShowNewError(IRedBoxErrorInfo info, RedBoxErrorType /*type*/) {
```

Comment:
Should probably just have the TestHostHarness implement the RedboxHandler instead of having the separate class.

Reply:
Actually I think that would add a reference cycle. Will keep as is.